File: patterns/java/src/refactoring_guru/patterns/iterator/example/LinkedInIterator.java
package refactoring_guru.patterns.iterator.example;

public class LinkedInIterator implements ProfileIterator {
    private LinkedIn linkedIn;
    private int profileId;
    private String type;
    private int currentPosition;
    private Profile[] cache;

    public LinkedInIterator(LinkedIn linkedIn, int profileId, String type) {
        this.linkedIn = linkedIn;
        this.profileId = profileId;
        this.type = type;
    }

    private void initIfNeeded() {
        if (cache == null) {
            cache = linkedIn.sendSophisticatedSocialGraphRequest(profileId);
        }
    }

    @Override
    public boolean hasNext() {
        initIfNeeded();
        return cache.length > currentPosition;
    }

    @Override
    public Profile getNext() {
        if (hasNext()) {
            currentPosition++;
            return cache[currentPosition];
        }
        return null;
    }
}

File: patterns/cs/src/RefactoringGuru/Iterator/Example/Iterators/LinkedInIterator.cs
using System;
using System.Collections.Generic;
using RefactoringGuru.Iterator.Example.Profiles;
using RefactoringGuru.Iterator.Example.SocialNetworks;

namespace RefactoringGuru.Iterator.Example.Iterators
{
    class LinkedInIterator : ProfileIterator
    {
        private LinkedIn linkedIn;
        private String type;
        private String email;
        private int currentPosition = 0;
        private List<String> emails = new List<string>();
        private List<Profile> contacts = new List<Profile>();

        public LinkedInIterator(LinkedIn linkedIn, String type, String email)
        {
            this.linkedIn = linkedIn;
            this.type = type;
            this.email = email;
        }

        private void LazyLoad()
        {
            if (emails.Count == 0)
            {
                List<String> profiles = linkedIn.RequestRelatedContactsFromLinkedInAPI(this.email, this.type);
                foreach (var profile in profiles)
                {
                    this.emails.Add(profile);
                    this.contacts.Add(null);
                }
            }
        }

        public bool HasNext()
        {
            LazyLoad();
            return currentPosition < emails.Count;
        }

        public Profile GetNext()
        {
            if (!HasNext())
            {
                return null;
            }

            String friendEmail = emails[currentPosition];
            Profile friendContact = contacts[currentPosition];
            if (friendContact == null)
            {
                friendContact = linkedIn.RequestContactInfoFromLinkedInAPI(friendEmail);
                contacts.Insert(currentPosition, friendContact);
            }
            currentPosition++;
            return friendContact;
        }

        public void Reset()
        {
            currentPosition = 0;
        }
    }
}
